Sources/SmokeDynamoDB/InMemoryDynamoDBCompositePrimaryKeyTableWithIndex.swift

```diff
@@ -19,334 +19,309 @@
+        self.primaryTable = InMemoryDynamoDBCompositePrimaryKeyTable(executeItemFilter: executeItemFilter)
+        self.gsiDataStore = InMemoryDynamoDBCompositePrimaryKeyTable(executeItemFilter: executeItemFilter)
     }
     
-    public func insertItem<AttributesType, ItemType>(_ item: TypedDatabaseItem<AttributesType, ItemType>)
-    -> EventLoopFuture<Void> where AttributesType : PrimaryKeyAttributes, ItemType : Decodable, ItemType : Encodable {
-        return self.primaryTable.insertItem(item) .flatMap { _ in
-            return self.gsiLogic.onInsertItem(item, gsiDataStore: self.gsiDataStore)
-        }
+    public func insertItem<AttributesType, ItemType>(_ item: TypedDatabaseItem<AttributesType, ItemType>) async throws {
+        try await self.primaryTable.insertItem(item)
+        try await self.gsiLogic.onInsertItem(item, gsiDataStore: self.gsiDataStore)
     }
     
-    public func clobberItem<AttributesType, ItemType>(_ item: TypedDatabaseItem<AttributesType, ItemType>)
-    -> EventLoopFuture<Void> where AttributesType : PrimaryKeyAttributes, ItemType : Decodable, ItemType : Encodable {
-        return self.primaryTable.clobberItem(item) .flatMap { _ in
-            return self.gsiLogic.onClobberItem(item, gsiDataStore: self.gsiDataStore)
-        }
+    public func clobberItem<AttributesType, ItemType>(_ item: TypedDatabaseItem<AttributesType, ItemType>) async throws {
+        try await self.primaryTable.clobberItem(item)
+        try await self.gsiLogic.onClobberItem(item, gsiDataStore: self.gsiDataStore)
     }
     
-    public func updateItem<AttributesType, ItemType>(newItem: TypedDatabaseItem<AttributesType, ItemType>, existingItem: TypedDatabaseItem<AttributesType, ItemType>)
-    -> EventLoopFuture<Void> where AttributesType : PrimaryKeyAttributes, ItemType : Decodable, ItemType : Encodable {
-        return self.primaryTable.updateItem(newItem: newItem, existingItem: existingItem) .flatMap { _ in
-            return self.gsiLogic.onUpdateItem(newItem: newItem, existingItem: existingItem, gsiDataStore: self.gsiDataStore)
-        }
+    public func updateItem<AttributesType, ItemType>(newItem: TypedDatabaseItem<AttributesType, ItemType>,
+                                                     existingItem: TypedDatabaseItem<AttributesType, ItemType>) async throws {
+        try await self.primaryTable.updateItem(newItem: newItem, existingItem: existingItem)
+        try await self.gsiLogic.onUpdateItem(newItem: newItem, existingItem: existingItem, gsiDataStore: self.gsiDataStore)
     }
     
-    public func monomorphicBulkWrite<AttributesType, ItemType>(_ entries: [WriteEntry<AttributesType, ItemType>])
-    -> EventLoopFuture<Void> {
-        let futures = entries.map { entry -> EventLoopFuture<Void> in
+    public func monomorphicBulkWrite<AttributesType, ItemType>(_ entries: [WriteEntry<AttributesType, ItemType>]) async throws {
+        try await entries.asyncForEach { entry in
```

Comment:
This will be a major version bump. The reason why I went for `asyncForEach` rather than `concurrentForEach` was because this is the in-memory implementation. It isn't actually making service calls so there should be any benefit to parallelising each entry.